### PR TITLE
Ensure order state handles with partial shipment

### LINKED ISSUE #38547: Order is not complete when mixing virtual, refunded and shipped items
### Preconditions and environment

- Magento 2.4.4-p7

### Steps to reproduce

1. Create an order containing: 1 physical item of quantity = 2 and 1 virtual item of quantity = 5
2. Invoice the order
3. Refund 1 quantity of the physical item
4. Ship 1 quantity of the physical item

image

### Expected result

Order state should be complete.

### Actual result

Order state is still processing.
image

### Additional information

Problem comes from `magento/module-sales/Model/ResourceModel/Order/Handler/State.php` 

In the `isPartiallyRefundedOrderShipped`, the code compares the total qty ordered with the sum of total qty shipped and total qty refunded.

However, virtual items are not considered when retrieving the total qty shipped, preventing the method from returning true.

Here's the fix that I have used to resolve the issue:

```
diff --git a/src/vendor/magento/module-sales/Model/ResourceModel/Order/Handler/State.php b/src/vendor/magento/module-sales/Model/ResourceModel/Order/Handler/State.php
index 51c45ed5e5..8d97ce01cb 100644
--- module-sales/Model/ResourceModel/Order/Handler/State.php
+++ module-sales/Model/ResourceModel/Order/Handler/State.php
@@ -97,1 +97,3 @@ class State
-            $numOfShippedItems += (int)$item->getQtyShipped();
+            $numOfShippedItems += $item->getIsVirtual()
+                ? (int)$item->getQtyOrdered()
+                : (int)$item->getQtyShipped();
```

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by digitalrisedorset**
The order flow will set the order to complete only once the order items are dealt with shipment, refund, cancellation. Yet, some edge case with virtual items and with partial shipment, partial refund, partial cancellation can make this flow faulty. 

### Description (*)
The change in this PR makes the `qty to ship` calculation to remove the qty refunded, qty cancelled and qty shipped already

### Fixed Issues (if relevant)

1. Fixes magento/magento2#38547

### Manual testing scenarios (*)
**Test 1:** place an order with only physical items
validate the order gets complete when the invoice and shipment are processed and all the items are shipped
**Test 2:** place an order with only physical items and partial refund
validate the order gets complete when the invoice, refund and shipment are processed and all the items are shipped or refund or cancelled
**Test 3:** place an order with a mix of configurable, physical items and virtual items
validate the order gets complete when the invoice, refund and shipment are processed and all the items are shipped or refund or cancelled

### Questions or comments
This change is not backward compliant. However, like @digitalpianism mentions, the current behaviour is not consistent when the order has only physical items. The latest change in the model `\Magento\Sales\Model\ResourceModel\Order\Handler\State` is what makes the experience consistent as per the requirements of the issue we are dealing with.

Currently, it is possible to refund an item that has not been shipped or just as well refund an item that has already been shipped without being able to trace either scenarios. Yet, the logic in the method `\Magento\Sales\Model\ResourceModel\Order\Handler\State::isPartiallyRefundedOrderShipped` assumes the refund and the shipping are separate information that can be cumulated

I have added some changes in the method` \Magento\Sales\Model\Order\Item::getSimpleQtyToShip`. These turn out to do the same as before. However, I would like to encourage to keep these changes (I am not precious but think they improve the readability of the logic in this method)

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by digitalrisedorset** [comment]
@magento run all tests

**Review by Den4ik — COMMENTED** [review]
@digitalrisedorset Thanks for contribution.
As I see suggested logic provide non expected logic.

Refund works only with shipped items.
Cancelation works only with non shipped items.

For example,
ordered = 5
shipped = 3
refunded = 2 (refunded from shipped 3)
canceled = 1
I expect to see 1 to ship.

Suggested logic will cause negative value and `max(round($qty, 8), 0)` return 0.

**Comment by digitalrisedorset** [comment]
excellent feedback, I will investigate this further. What I am taking from your comment is that the refund can be either done on an item that has not been shipped or alternatively on an item that has been shipped.
In both scenarios, the calculations to decide whether some items are yet to be shipped is different. 
I will try to understand whether we can define if an item has been refund before the shipment or after and then if we can have a way to pinpoint this, I will adjust the calculation to fit both scenarios. 

Many thanks @Den4ik

**Comment by digitalrisedorset** [comment]
@Den4ik following your inputs, I was able to identify more use cases:

**Use case 1: the example you gave**
ordered = 5
shipped = 3
refunded = 2 (refunded from shipped 3)
canceled = 1
I expect to see 1 to ship.

**Use case 2: order 5 physical products: simplest use case**
ordered 5 
shipped 5
refund 0
cancelled 0
→ qty to ship 0


**Use case 3: order 5 physical products, ship 5 and refund 2** 
ordered 5 
shipped 5
refund 2
cancelled 0
→ qty to ship 0

**Use case 4: order 5 physical products, ship 2, refund 2**
ordered 5 
shipped 2
refund 2
cancelled 0
→ qty to ship 3

**Use case 5: order 5 with 2 shipped and physical products with 2 refunded before shipment and 1 cancelled**
ordered 5 
shipped 2
refund 4
cancelled 0
→ qty to ship 1

The additional thoughts that have come to my attention are:
- currently, the unit tests are written in a way to implies that we can ship even if the order is not paid (in reality, only the invoiced items should be eligible to be shipped)
- the result of getQtyInvoiced and getQtyShipped is a float such as 0.0 and yet the codebase interact with the figure 0 at times
- ultimately, the big finding I made is that if the refund qty is less than the shipping, then these refund data do not change the qty to ship. If the qty refunded is higher than the qty shipped, then I substract the qty refund and qty shipped

**Comment by digitalrisedorset** [comment]
@Den4ik I have now found out a way to fix the problem and I have committed some changes today. However, I see my change has a conflict with the method` \Magento\Sales\Model\ResourceModel\Order\Handler\State::isPartiallyRefundedOrderShipped`

The 2 lines change that have been made and conflict with my change are sufficient for this issue to be called close. I am more than happy to go with the change that have been provided and discard all my changes. 

**In short, if we close my PR and the issue and ensure the lines that conflict with my changes get pushed further, then the issue is resolved**

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

Have you had a chance to look into this https://github.com/magento/magento2/pull/38679#issuecomment-2473420015?

Thanks

**Comment by digitalrisedorset** [comment]
> Hello @digitalrisedorset,
> 
> Have you had a chance to look into this [#38679 (comment)](https://github.com/magento/magento2/pull/38679#issuecomment-2473420015)?
> 
> Thanks

Hi, I have not looked into it. My PRs are 6 months old and I had no feedback for all this time. I have now moved on. Apologies.

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

Apologies for the delay. Currently, we are focusing on PRs from the `Community Picked` dashboard. Since your PR has been voted as `Community Picked`, we have selected it for further processing.

We kindly request you to resolve the conflicts so we can proceed with this PR.

Thank you for your contribution!

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

A gentle reminder to resolve the conflicts on this PR.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

Have you got a chance to fix the conflicts on this PR?

Thanks

**Comment by digitalrisedorset** [comment]
Hi Abhinav,

I plan to take a look when I am back from the christmas break. 
Apologies, the end of year has been challenging and I have not been able 
to take a look at the conflict.

Kind regards,

Herve

On 23/12/2024 13:49, Abhinav Pathak wrote:

> Hello @digitalrisedorset [1],
> 
> Have you got a chance to fix the conflicts on this PR?
> 
> Thanks
> 
> --
> Reply to this email directly, view it on GitHub [2], or unsubscribe 
> [3].
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>


Links:
------
[1] https://github.com/digitalrisedorset
[2] 
https://github.com/magento/magento2/pull/38679#issuecomment-2559743861
[3]

**Comment by engcom-Hotel** [comment]
Sure @digitalrisedoreset,

Please resolve the conflicts whenever you are back.

Thanks

**Comment by digitalrisedorset** [comment]
> Sure @digitalrisedoreset,
> 
> Please resolve the conflicts whenever you are back.
> 
> Thanks

I have made some changes and submitted a PR now. Apologies, I was hoping to do better than this. The  steps I have taken are as follows:
- I pulled the latests develop branch
- I created a new branch and integrated my changes for the issue we are working onto. 
- The caveats with my change is that the tests side is not changed. I have not been working on Magento projects for a while and I know the tests have changed a lot in the last years so I thought you will be doing this part better than me.

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

We are still seeing the conflicts on this PR. Have you created a new PR for the same? If yes can you please provide the details of that PR.

Thanks

**Comment by digitalrisedorset** [comment]
The new branch that I have created is https://github.com/digitalrisedorset/magento2/tree/feature/order-state-faulty-with-partial-shipment-with-conflict-resolved

**Comment by engcom-Hotel** [comment]
Thanks @digitalrisedorset, but we are asking about the PR, if you have created any. Because on this PR we can see the conflicts are still there.

**Comment by digitalrisedorset** [comment]
Apologies team, I think it might be worth to organise a meeting. I have created most of my PRs 10 months ago, I have to say it was disappointing not to have more follow-up earlier. Today, I am torn as to whether my inputs will be taken into account. My suggestion is to work closely to ensure my time is not wasted and rather we set good foundations to work on a good roadmap to help each other.

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

Pinged you over slack. Lets discuss there.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

It was great discussing with you on Slack yesterday. As per our discussion, we have a new PR https://github.com/magento/magento2/pull/39548, created by @digitalrisedorset, with resolved conflicts. Therefore, we can proceed with this PR.

As a result, we are closing this PR.

Thank you

**Comment by Den4ik** [comment]
@engcom-Hotel Thanks for update

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

As per our discussion over slack, we will continue with this PR for resolving the conflicts. But it seems we dont have necessary permission to push the code on this branch. Please allow us the same.

We are getting the below error while pushing the code:

image

Please do the needful. Meanwhile moving this PR `On Hold`.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @digitalrisedorset,

Have you got a chance to look into this https://github.com/magento/magento2/pull/38679#issuecomment-2714161728? We are unable to push code due to access issue. Please provide us the required permission.

Thanks

**Comment by Den4ik** [comment]
@digitalrisedorset Will you have a chance resolve conflicts?

**Comment by digitalrisedorset** [comment]
HI, I am happy to give you the permissions. Shall we meet to validate the permissions you need. If you ping me a date and time when we can meet. I'd like as well to clear this for you to benefit of the PR.

**Comment by Den4ik** [comment]
@digitalrisedorset Please follow this instruction https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork
